validate/validate.go

```diff
@@ -0,0 +1,574 @@
+package validate
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"unicode"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/blang/semver"
+	rspec "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+var (
+	defaultRlimits = []string{
+		"RLIMIT_CPU",
+		"RLIMIT_FSIZE",
+		"RLIMIT_DATA",
+		"RLIMIT_STACK",
+		"RLIMIT_CORE",
+		"RLIMIT_RSS",
+		"RLIMIT_NPROC",
+		"RLIMIT_NOFILE",
+		"RLIMIT_MEMLOCK",
+		"RLIMIT_AS",
+		"RLIMIT_LOCKS",
+		"RLIMIT_SIGPENDING",
+		"RLIMIT_MSGQUEUE",
+		"RLIMIT_NICE",
+		"RLIMIT_RTPRIO",
+		"RLIMIT_RTTIME",
+	}
+	defaultCaps = []string{
+		"CAP_CHOWN",
+		"CAP_DAC_OVERRIDE",
+		"CAP_FSETID",
+		"CAP_FOWNER",
+		"CAP_MKNOD",
+		"CAP_NET_RAW",
+		"CAP_SETGID",
+		"CAP_SETUID",
+		"CAP_SETFCAP",
+		"CAP_SETPCAP",
+		"CAP_NET_BIND_SERVICE",
+		"CAP_SYS_CHROOT",
+		"CAP_KILL",
+		"CAP_AUDIT_WRITE",
+	}
+)
+
+type Validator struct {
+	spec         rspec.Spec
+	rootfs       string
+	hostSpecific bool
```

Comment:
On Wed, Sep 07, 2016 at 09:04:44PM -0700, 梁辰晔 (Liang Chenye) wrote:
> > +type Validator struct {
> > +	spec         rspec.Spec
> > +	rootfs       string
> > +	hostSpecific bool
>
> Do we need to see and modify these? I think the 'Check*' functions
> are enough.

Making them public means you don't need getters/setters and you can
still support users who want to partially setup the instance when they
call NewValidator (or whatever [1]) and finish the setup later.  Or if
they want to peer inside to get more details on why their validation
is failing.  Or who knows?  I'd rather err on the side of “we're all
consenting adults” [2], because I don't see a way to make validation
go terribly wrong if a user gets or sets one of these values whenever
they like.

[1]: https://github.com/opencontainers/ocitools/pull/188/files#r77942970
[2]: https://github.com/opencontainers/ocitools/pull/138#discussion_r71917014